**Linux** [Build 8129](http://brontes.lab.eng.brq.redhat.com/viewLog.html?buildId=70764&buildTypeId=WFPR) outcome was **FAILURE** using a merge of a0084a55c5bb682d23ee6c24ec93c6f395f1febc
Summary: Tests failed: 3 (3 new), passed: 3119, ignored: 364 Build time: 00:50:42
### Failed tests
```
org.jboss.as.test.integration.security.jaspi.JASPIHttpSchemeServerAuthModelTestCase.testRequiresAuthentication: java.lang.AssertionError: expected:<401> but was:<200>
	at org.jboss.as.test.integration.security.jaspi.JASPIHttpSchemeServerAuthModelTestCase.testRequiresAuthentication(JASPIHttpSchemeServerAuthModelTestCase.java:77)


org.jboss.as.test.integration.security.jaspi.JASPIHttpSchemeServerAuthModelTestCase.testSuccessfulAuthentication: java.lang.AssertionError: null
	at org.jboss.as.test.integration.security.jaspi.JASPIHttpSchemeServerAuthModelTestCase.testSuccessfulAuthentication(JASPIHttpSchemeServerAuthModelTestCase.java:107)


org.jboss.as.test.integration.security.jaspi.JASPIHttpSchemeServerAuthModelTestCase.testUnsuccessfulAuthentication: java.lang.AssertionError: expected:<401> but was:<200>
	at org.jboss.as.test.integration.security.jaspi.JASPIHttpSchemeServerAuthModelTestCase.testUnsuccessfulAuthentication(JASPIHttpSchemeServerAuthModelTestCase.java:116)


```

--- a/undertow/src/main/java/org/wildfly/extension/undertow/deployment/UndertowDeploymentInfoService.java
+++ b/undertow/src/main/java/org/wildfly/extension/undertow/deployment/UndertowDeploymentInfoService.java
@@ -149,6 +149,7 @@ import org.wildfly.extension.undertow.security.SecurityContextThreadSetupAction;
 import org.wildfly.extension.undertow.security.jacc.JACCAuthorizationManager;
 import org.wildfly.extension.undertow.security.jacc.JACCContextIdHandler;
 import org.wildfly.extension.undertow.security.jaspi.JASPIAuthenticationMechanism;
+import org.wildfly.extension.undertow.security.jaspi.JASPICInitialHandler;
 import org.wildfly.extension.undertow.security.jaspi.JASPICSecurityContextFactory;
 import org.wildfly.extension.undertow.session.CodecSessionConfigWrapper;
 import org.wildfly.extension.undertow.session.SharedSessionManagerConfig;
@@ -461,8 +462,13 @@ public class UndertowDeploymentInfoService implements Service<DeploymentInfo> {
             LoginConfig loginConfig = deploymentInfo.getLoginConfig();
             if (loginConfig != null && loginConfig.getAuthMethods().size() > 0)
                 authMethod = loginConfig.getAuthMethods().get(0).getName();
-
-            deploymentInfo.setJaspiAuthenticationMechanism(new JASPIAuthenticationMechanism(this.securityDomain, authMethod));
+            deploymentInfo.addSecurityWrapper(new HandlerWrapper() {
+                @Override
+                public HttpHandler wrap(HttpHandler handler) {
+                    return new JASPICInitialHandler(securityDomain, handler);
+                }
+            });
+            deploymentInfo.setJaspiAuthenticationMechanism(new JASPIAuthenticationMechanism(authMethod));
             deploymentInfo.setSecurityContextFactory(new JASPICSecurityContextFactory(this.securityDomain));
         }
     }

--- a/undertow/src/main/java/org/wildfly/extension/undertow/security/jaspi/JASPIAuthenticationMechanism.java
+++ b/undertow/src/main/java/org/wildfly/extension/undertow/security/jaspi/JASPIAuthenticationMechanism.java
@@ -21,30 +21,20 @@
  */
 package org.wildfly.extension.undertow.security.jaspi;
 
-import io.undertow.security.api.AuthenticatedSessionManager;
 import io.undertow.security.api.AuthenticationMechanism;
 import io.undertow.security.api.SecurityContext;
 import io.undertow.security.idm.Account;
-import io.undertow.server.ConduitWrapper;
 import io.undertow.server.HttpServerExchange;
 import io.undertow.servlet.handlers.ServletRequestContext;
 import io.undertow.util.AttachmentKey;
 
-import io.undertow.util.ConduitFactory;
 import org.jboss.security.SecurityConstants;
-import org.jboss.security.auth.callback.JBossCallbackHandler;
 import org.jboss.security.auth.message.GenericMessageInfo;
 import org.jboss.security.identity.plugins.SimpleRole;
 import org.jboss.security.identity.plugins.SimpleRoleGroup;
 import org.jboss.security.plugins.auth.JASPIServerAuthenticationManager;
 import org.wildfly.extension.undertow.logging.UndertowLogger;
 import org.wildfly.extension.undertow.security.AccountImpl;
-
-import javax.security.auth.Subject;
-import javax.security.auth.message.AuthException;
-import javax.servlet.ServletRequest;
-import javax.servlet.http.HttpServletRequest;
-import javax.servlet.http.HttpServletResponse;
 
 import java.security.Principal;
 import java.util.HashSet;
@@ -53,8 +43,6 @@ import java.util.Set;
 import org.jboss.security.auth.callback.JASPICallbackHandler;
 import org.jboss.security.identity.Role;
 import org.jboss.security.identity.RoleGroup;
-import org.wildfly.extension.undertow.security.UndertowSecurityAttachments;
-import org.xnio.conduits.StreamSinkConduit;
 
 /**
  * <p>
@@ -73,52 +61,30 @@ public class JASPIAuthenticationMechanism implements AuthenticationMechanism {
 
     public static final AttachmentKey<HttpServerExchange> HTTP_SERVER_EXCHANGE_ATTACHMENT_KEY = AttachmentKey.create(HttpServerExchange.class);
     public static final AttachmentKey<SecurityContext> SECURITY_CONTEXT_ATTACHMENT_KEY = AttachmentKey.create(SecurityContext.class);
-    private static final AttachmentKey<Boolean> ALREADY_WRAPPED = AttachmentKey.create(Boolean.class);
 
-    private final String securityDomain;
     private final String configuredAuthMethod;
 
-    public JASPIAuthenticationMechanism(final String securityDomain, final String configuredAuthMethod) {
-        this.securityDomain = securityDomain;
+    public JASPIAuthenticationMechanism(final String configuredAuthMethod) {
         this.configuredAuthMethod = configuredAuthMethod;
     }
 
     @Override
     public AuthenticationMechanismOutcome authenticate(final HttpServerExchange exchange, final SecurityContext sc) {
-        final ServletRequestContext requestContext = exchange.getAttachment(ServletRequestContext.ATTACHMENT_KEY);
-        final JASPIServerAuthenticationManager sam = createJASPIAuthenticationManager();
-        final GenericMessageInfo messageInfo = createMessageInfo(exchange, sc);
-        final String applicationIdentifier = buildApplicationIdentifier(requestContext);
-        final JASPICallbackHandler cbh = new JASPICallbackHandler();
+        JASPICAttachment attachment = exchange.getAttachment(JASPICAttachment.ATTACHMENT_KEY);
 
-        UndertowLogger.ROOT_LOGGER.debugf("validateRequest for layer [%s] and applicationContextIdentifier [%s]", JASPI_HTTP_SERVLET_LAYER, applicationIdentifier);
-
-        Account cachedAccount = null;
-        final JASPICSecurityContext jaspicSecurityContext = (JASPICSecurityContext) exchange.getSecurityContext();
-        final AuthenticatedSessionManager sessionManager = exchange.getAttachment(AuthenticatedSessionManager.ATTACHMENT_KEY);
-
-        if (sessionManager != null) {
-            AuthenticatedSessionManager.AuthenticatedSession authSession = sessionManager.lookupSession(exchange);
-            if(authSession != null) {
-                cachedAccount = authSession.getAccount();
-                // if there is a cached account we set it in the security context so that the principal is available to
-                // SAM modules via request.getUserPrincipal().
-                if (cachedAccount != null) {
-                    jaspicSecurityContext.setCachedAuthenticatedAccount(cachedAccount);
-                }
-            }
-        }
-
-        AuthenticationMechanismOutcome outcome = AuthenticationMechanismOutcome.NOT_AUTHENTICATED;
+        AuthenticationMechanismOutcome outcome;
         Account authenticatedAccount = null;
 
-        boolean isValid = sam.isValid(messageInfo, new Subject(), JASPI_HTTP_SERVLET_LAYER, applicationIdentifier, cbh);
-        jaspicSecurityContext.setCachedAuthenticatedAccount(null);
+        boolean isValid = attachment.isValid();
+        final ServletRequestContext requestContext = attachment.getRequestContext();
+        final JASPIServerAuthenticationManager sam = attachment.getSam();
+        final JASPICallbackHandler cbh = attachment.getCbh();
 
+        GenericMessageInfo messageInfo = attachment.getMessageInfo();
         if (isValid) {
             // The CBH filled in the JBOSS SecurityContext, we need to create an Undertow account based on that
             org.jboss.security.SecurityContext jbossSct = SecurityActions.getSecurityContext();
-            authenticatedAccount = createAccount(cachedAccount, jbossSct);
+            authenticatedAccount = createAccount(attachment.getCachedAccount(), jbossSct);
         }
 
         // authType resolution (check message info first, then check for the configured auth method, then use mech-specific name).
@@ -142,13 +108,6 @@ public class JASPIAuthenticationMechanism implements AuthenticationMechanism {
             sc.authenticationFailed("JASPIC authentication failed.", authType);
         }
 
-        // A SAM can wrap the HTTP request/response objects - update the servlet request context with the values found in the message info.
-        ServletRequestContext servletRequestContext = exchange.getAttachment(ServletRequestContext.ATTACHMENT_KEY);
-        servletRequestContext.setServletRequest((HttpServletRequest) messageInfo.getRequestMessage());
-        servletRequestContext.setServletResponse((HttpServletResponse) messageInfo.getResponseMessage());
-
-        secureResponse(exchange, sam, messageInfo, cbh);
-
         return outcome;
 
     }
@@ -156,36 +115,6 @@ public class JASPIAuthenticationMechanism implements AuthenticationMechanism {
     @Override
     public ChallengeResult sendChallenge(final HttpServerExchange exchange, final SecurityContext securityContext) {
         return new ChallengeResult(true);
-    }
-
-    private boolean wasAuthExceptionThrown(HttpServerExchange exchange) {
-        return exchange.getAttachment(UndertowSecurityAttachments.SECURITY_CONTEXT_ATTACHMENT).getData().get(AuthException.class.getName()) != null;
-    }
-
-    private JASPIServerAuthenticationManager createJASPIAuthenticationManager() {
-        return new JASPIServerAuthenticationManager(this.securityDomain, new JBossCallbackHandler());
-    }
-
-    private String buildApplicationIdentifier(final ServletRequestContext attachment) {
-        ServletRequest servletRequest = attachment.getServletRequest();
-        return servletRequest.getServletContext().getVirtualServerName() + " " + servletRequest.getServletContext().getContextPath();
-    }
-
-    private GenericMessageInfo createMessageInfo(final HttpServerExchange exchange, final SecurityContext securityContext) {
-        ServletRequestContext servletRequestContext = exchange.getAttachment(ServletRequestContext.ATTACHMENT_KEY);
-
-        GenericMessageInfo messageInfo = new GenericMessageInfo();
-
-        messageInfo.setRequestMessage(servletRequestContext.getServletRequest());
-        messageInfo.setResponseMessage(servletRequestContext.getServletResponse());
-
-        messageInfo.getMap().put("javax.security.auth.message.MessagePolicy.isMandatory", isMandatory(servletRequestContext).toString());
-
-        // additional context data, useful to provide access to Undertow resources during the modules processing
-        messageInfo.getMap().put(SECURITY_CONTEXT_ATTACHMENT_KEY, securityContext);
-        messageInfo.getMap().put(HTTP_SERVER_EXCHANGE_ATTACHMENT_KEY, exchange);
-
-        return messageInfo;
     }
 
     private Account createAccount(final Account cachedAccount, final org.jboss.security.SecurityContext jbossSct) {
@@ -224,32 +153,6 @@ public class JASPIAuthenticationMechanism implements AuthenticationMechanism {
             original = cachedAccount.getPrincipal();
         }
         return new AccountImpl(userPrincipal, stringRoles, credential, original);
-    }
-
-    private void secureResponse(final HttpServerExchange exchange, final JASPIServerAuthenticationManager sam, final GenericMessageInfo messageInfo, final JASPICallbackHandler cbh) {
-        if(exchange.getAttachment(ALREADY_WRAPPED) != null || exchange.isResponseStarted()) {
-            return;
-        }
-        exchange.putAttachment(ALREADY_WRAPPED, true);
-        // we add a response wrapper to properly invoke the secureResponse, after processing the destination
-        exchange.addResponseWrapper(new ConduitWrapper<StreamSinkConduit>() {
-            @Override
-            public StreamSinkConduit wrap(final ConduitFactory<StreamSinkConduit> factory, final HttpServerExchange exchange) {
-                ServletRequestContext requestContext = exchange.getAttachment(ServletRequestContext.ATTACHMENT_KEY);
-                String applicationIdentifier = buildApplicationIdentifier(requestContext);
-
-                if (!wasAuthExceptionThrown(exchange)) {
-                    UndertowLogger.ROOT_LOGGER.debugf("secureResponse for layer [%s] and applicationContextIdentifier [%s].", JASPI_HTTP_SERVLET_LAYER, applicationIdentifier);
-                    sam.secureResponse(messageInfo, new Subject(), JASPI_HTTP_SERVLET_LAYER, applicationIdentifier, cbh);
-
-                    // A SAM can unwrap the HTTP request/response objects - update the servlet request context with the values found in the message info.
-                    ServletRequestContext servletRequestContext = exchange.getAttachment(ServletRequestContext.ATTACHMENT_KEY);
-                    servletRequestContext.setServletRequest((HttpServletRequest) messageInfo.getRequestMessage());
-                    servletRequestContext.setServletResponse((HttpServletResponse) messageInfo.getResponseMessage());
-                }
-                return factory.create();
-            }
-        });
     }
 
     /**

--- a/undertow/src/main/java/org/wildfly/extension/undertow/security/jaspi/JASPICAttachment.java
+++ b/undertow/src/main/java/org/wildfly/extension/undertow/security/jaspi/JASPICAttachment.java
@@ -1,0 +1,62 @@
+package org.wildfly.extension.undertow.security.jaspi;
+
+import io.undertow.security.idm.Account;
+import io.undertow.servlet.handlers.ServletRequestContext;
+import io.undertow.util.AttachmentKey;
+import org.jboss.security.auth.callback.JASPICallbackHandler;
+import org.jboss.security.auth.message.GenericMessageInfo;
+import org.jboss.security.plugins.auth.JASPIServerAuthenticationManager;
+
+/**
+ * @author Stuart Douglas
+ */
+class JASPICAttachment {
+
+    public static final AttachmentKey<JASPICAttachment> ATTACHMENT_KEY = AttachmentKey.create(JASPICAttachment.class);
+
+    private final boolean isValid;
+    private final ServletRequestContext requestContext;
+    private final JASPIServerAuthenticationManager sam;
+    private final GenericMessageInfo messageInfo;
+    private final String applicationIdentifier;
+    private final JASPICallbackHandler cbh;
+    private final Account cachedAccount;
+
+    JASPICAttachment(boolean isValid, ServletRequestContext requestContext, JASPIServerAuthenticationManager sam, GenericMessageInfo messageInfo, String applicationIdentifier, JASPICallbackHandler cbh, Account cachedAccount) {
+        this.isValid = isValid;
+        this.requestContext = requestContext;
+        this.sam = sam;
+        this.messageInfo = messageInfo;
+        this.applicationIdentifier = applicationIdentifier;
+        this.cbh = cbh;
+        this.cachedAccount = cachedAccount;
+    }
+
+    public boolean isValid() {
+        return isValid;
+    }
+
+    public ServletRequestContext getRequestContext() {
+        return requestContext;
+    }
+
+    public JASPIServerAuthenticationManager getSam() {
+        return sam;
+    }
+
+    public GenericMessageInfo getMessageInfo() {
+        return messageInfo;
+    }
+
+    public String getApplicationIdentifier() {
+        return applicationIdentifier;
+    }
+
+    public JASPICallbackHandler getCbh() {
+        return cbh;
+    }
+
+    public Account getCachedAccount() {
+        return cachedAccount;
+    }
+}

--- a/undertow/src/main/java/org/wildfly/extension/undertow/security/jaspi/JASPICInitialHandler.java
+++ b/undertow/src/main/java/org/wildfly/extension/undertow/security/jaspi/JASPICInitialHandler.java
@@ -1,0 +1,180 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2013, Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.wildfly.extension.undertow.security.jaspi;
+
+import io.undertow.security.api.AuthenticatedSessionManager;
+import io.undertow.security.api.SecurityContext;
+import io.undertow.security.idm.Account;
+import io.undertow.server.ConduitWrapper;
+import io.undertow.server.HttpHandler;
+import io.undertow.server.HttpServerExchange;
+import io.undertow.servlet.handlers.ServletRequestContext;
+import io.undertow.util.AttachmentKey;
+import io.undertow.util.ConduitFactory;
+import org.jboss.security.auth.callback.JASPICallbackHandler;
+import org.jboss.security.auth.callback.JBossCallbackHandler;
+import org.jboss.security.auth.message.GenericMessageInfo;
+import org.jboss.security.plugins.auth.JASPIServerAuthenticationManager;
+import org.wildfly.extension.undertow.logging.UndertowLogger;
+import org.wildfly.extension.undertow.security.UndertowSecurityAttachments;
+import org.xnio.conduits.StreamSinkConduit;
+
+import javax.security.auth.Subject;
+import javax.security.auth.message.AuthException;
+import javax.servlet.ServletRequest;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * <p>
+ * {@link io.undertow.security.api.AuthenticationMechanism} implementation that enables JASPI-based authentication.
+ * </p>
+ *
+ * @author Stuart Douglas
+ * @author Pedro Igor
+ * @author <a href="mailto:sguilhen@redhat.com">Stefan Guilhen</a>
+ */
+public class JASPICInitialHandler implements HttpHandler {
+
+    private static final String JASPI_HTTP_SERVLET_LAYER = "HttpServlet";
+
+    private static final AttachmentKey<Boolean> ALREADY_WRAPPED = AttachmentKey.create(Boolean.class);
+
+    private final String securityDomain;
+    private final HttpHandler next;
+
+    public JASPICInitialHandler(final String securityDomain, HttpHandler next) {
+        this.securityDomain = securityDomain;
+        this.next = next;
+    }
+
+    @Override
+    public void handleRequest(HttpServerExchange exchange) throws Exception {
+        final ServletRequestContext requestContext = exchange.getAttachment(ServletRequestContext.ATTACHMENT_KEY);
+        final JASPIServerAuthenticationManager sam = createJASPIAuthenticationManager();
+        final GenericMessageInfo messageInfo = createMessageInfo(exchange, exchange.getSecurityContext());
+        final String applicationIdentifier = buildApplicationIdentifier(requestContext);
+        final JASPICallbackHandler cbh = new JASPICallbackHandler();
+
+        UndertowLogger.ROOT_LOGGER.debugf("validateRequest for layer [%s] and applicationContextIdentifier [%s]", JASPI_HTTP_SERVLET_LAYER, applicationIdentifier);
+
+        Account cachedAccount = null;
+        final JASPICSecurityContext jaspicSecurityContext = (JASPICSecurityContext) exchange.getSecurityContext();
+        final AuthenticatedSessionManager sessionManager = exchange.getAttachment(AuthenticatedSessionManager.ATTACHMENT_KEY);
+
+        if (sessionManager != null) {
+            AuthenticatedSessionManager.AuthenticatedSession authSession = sessionManager.lookupSession(exchange);
+            if(authSession != null) {
+                cachedAccount = authSession.getAccount();
+                // if there is a cached account we set it in the security context so that the principal is available to
+                // SAM modules via request.getUserPrincipal().
+                if (cachedAccount != null) {
+                    jaspicSecurityContext.setCachedAuthenticatedAccount(cachedAccount);
+                }
+            }
+        }
+
+        boolean isValid = sam.isValid(messageInfo, new Subject(), JASPI_HTTP_SERVLET_LAYER, applicationIdentifier, cbh);
+
+        jaspicSecurityContext.setCachedAuthenticatedAccount(null);
+
+        exchange.putAttachment(JASPICAttachment.ATTACHMENT_KEY, new JASPICAttachment(isValid, requestContext, sam, messageInfo, applicationIdentifier, cbh, cachedAccount));
+
+
+
+        // A SAM can wrap the HTTP request/response objects - update the servlet request context with the values found in the message info.
+        ServletRequestContext servletRequestContext = exchange.getAttachment(ServletRequestContext.ATTACHMENT_KEY);
+        servletRequestContext.setServletRequest((HttpServletRequest) messageInfo.getRequestMessage());
+        servletRequestContext.setServletResponse((HttpServletResponse) messageInfo.getResponseMessage());
+
+
+        secureResponse(exchange, sam, messageInfo, cbh);
+        next.handleRequest(exchange);
+    }
+
+    private JASPIServerAuthenticationManager createJASPIAuthenticationManager() {
+        return new JASPIServerAuthenticationManager(this.securityDomain, new JBossCallbackHandler());
+    }
+
+    private void secureResponse(final HttpServerExchange exchange, final JASPIServerAuthenticationManager sam, final GenericMessageInfo messageInfo, final JASPICallbackHandler cbh) {
+        if(exchange.getAttachment(ALREADY_WRAPPED) != null || exchange.isResponseStarted()) {
+            return;
+        }
+        exchange.putAttachment(ALREADY_WRAPPED, true);
+        // we add a response wrapper to properly invoke the secureResponse, after processing the destination
+        exchange.addResponseWrapper(new ConduitWrapper<StreamSinkConduit>() {
+            @Override
+            public StreamSinkConduit wrap(final ConduitFactory<StreamSinkConduit> factory, final HttpServerExchange exchange) {
+                ServletRequestContext requestContext = exchange.getAttachment(ServletRequestContext.ATTACHMENT_KEY);
+                String applicationIdentifier = buildApplicationIdentifier(requestContext);
+
+                if (!wasAuthExceptionThrown(exchange)) {
+                    UndertowLogger.ROOT_LOGGER.debugf("secureResponse for layer [%s] and applicationContextIdentifier [%s].", JASPI_HTTP_SERVLET_LAYER, applicationIdentifier);
+                    sam.secureResponse(messageInfo, new Subject(), JASPI_HTTP_SERVLET_LAYER, applicationIdentifier, cbh);
+
+                    // A SAM can unwrap the HTTP request/response objects - update the servlet request context with the values found in the message info.
+                    ServletRequestContext servletRequestContext = exchange.getAttachment(ServletRequestContext.ATTACHMENT_KEY);
+                    servletRequestContext.setServletRequest((HttpServletRequest) messageInfo.getRequestMessage());
+                    servletRequestContext.setServletResponse((HttpServletResponse) messageInfo.getResponseMessage());
+                }
+                return factory.create();
+            }
+        });
+    }
+
+    private boolean wasAuthExceptionThrown(HttpServerExchange exchange) {
+        return exchange.getAttachment(UndertowSecurityAttachments.SECURITY_CONTEXT_ATTACHMENT).getData().get(AuthException.class.getName()) != null;
+    }
+
+    private String buildApplicationIdentifier(final ServletRequestContext attachment) {
+        ServletRequest servletRequest = attachment.getServletRequest();
+        return servletRequest.getServletContext().getVirtualServerName() + " " + servletRequest.getServletContext().getContextPath();
+    }
+
+    private GenericMessageInfo createMessageInfo(final HttpServerExchange exchange, final SecurityContext securityContext) {
+        ServletRequestContext servletRequestContext = exchange.getAttachment(ServletRequestContext.ATTACHMENT_KEY);
+
+        GenericMessageInfo messageInfo = new GenericMessageInfo();
+
+        messageInfo.setRequestMessage(servletRequestContext.getServletRequest());
+        messageInfo.setResponseMessage(servletRequestContext.getServletResponse());
+
+        messageInfo.getMap().put("javax.security.auth.message.MessagePolicy.isMandatory", isMandatory(servletRequestContext).toString());
+
+        // additional context data, useful to provide access to Undertow resources during the modules processing
+        messageInfo.getMap().put(JASPIAuthenticationMechanism.SECURITY_CONTEXT_ATTACHMENT_KEY, securityContext);
+        messageInfo.getMap().put(JASPIAuthenticationMechanism.HTTP_SERVER_EXCHANGE_ATTACHMENT_KEY, exchange);
+
+        return messageInfo;
+    }
+
+    /**
+     * <p>The authentication is mandatory if the servlet has http constraints (eg.: {@link
+     * javax.servlet.annotation.HttpConstraint}).</p>
+     *
+     * @param attachment
+     * @return
+     */
+    private Boolean isMandatory(final ServletRequestContext attachment) {
+        return attachment.getExchange().getSecurityContext() != null && attachment.getExchange().getSecurityContext().isAuthenticationRequired();
+    }
+}

--- a/undertow/src/main/java/org/wildfly/extension/undertow/security/jaspi/modules/HTTPSchemeServerAuthModule.java
+++ b/undertow/src/main/java/org/wildfly/extension/undertow/security/jaspi/modules/HTTPSchemeServerAuthModule.java
@@ -93,17 +93,23 @@ public class HTTPSchemeServerAuthModule implements ServerAuthModule {
                     break;
                 }
             }
+
             if (!success) {
-                for (AuthenticationMechanism mechanism : mechanisms) {
-                    AuthenticationMechanism.ChallengeResult challengeResult = mechanism.sendChallenge(exchange, securityContext);
-                    if (challengeResult.getDesiredResponseCode() != null) {
-                        exchange.setResponseCode(challengeResult.getDesiredResponseCode());
+                String mandatory = (String) messageInfo.getMap().get("javax.security.auth.message.MessagePolicy.isMandatory");
+                if(mandatory != null && mandatory.toLowerCase().equals("false")) {
+                    return SUCCESS;
+                } else {
+                    for (AuthenticationMechanism mechanism : mechanisms) {
+                        AuthenticationMechanism.ChallengeResult challengeResult = mechanism.sendChallenge(exchange, securityContext);
+                        if (challengeResult.getDesiredResponseCode() != null) {
+                            exchange.setResponseCode(challengeResult.getDesiredResponseCode());
+                        }
+                        if (exchange.isResponseComplete()) {
+                            break;
+                        }
                     }
-                    if(exchange.isResponseComplete()) {
-                        break;
-                    }
+                    return SEND_CONTINUE;
                 }
-                return SEND_CONTINUE;
             }
         } catch (Exception e) {
             UndertowLogger.ROOT_LOGGER.debug(e);

